File: Adapter/Payment.js
// TARGET — what client expects
class OldPaymentMethod {
  makePayment(amount) {
    throw new Error('makePayment() must be implemented')
  }
}

// ADAPTEE — new gateway with different interface
// completely independent — no relationship to OldPaymentMethod
class NewPaymentGateway {
  processPayment(amt) {
    console.log(`✅ Payment of ₹${amt} done via new gateway`)
  }
}

// ADAPTER — extends target, wraps adaptee
class PaymentAdapter extends OldPaymentMethod {
  constructor() {
    super()
    this.newGateway = new NewPaymentGateway()  // wraps adaptee
  }

  makePayment(amount) {
    // translates makePayment() → processPayment()
    this.newGateway.processPayment(amount)
  }
}

// CLIENT — only knows OldPaymentMethod interface
const client = new PaymentAdapter()
client.makePayment(200)
// ✅ Payment of ₹200 done via new gateway

File: Adapter/Payment.java
package Adapter;

class OldPaymentMathod{ //Target
    public void makePayment(int amt){
        System.out.println(amt+" paid using old method");
    }
}
class NewPaymentMethod{ //Adaptee
    public void processPayment(int amt){
        System.out.println(amt+" paid using new method");
    }
}
class Adapter extends OldPaymentMathod{ //Adapter
    NewPaymentMethod gateway;
    Adapter(){
        this.gateway = new NewPaymentMethod();
    }
    public void makePayment(int amt){
        this.gateway.processPayment(amt);
    }
}
public class Payment {
    public static void main(String[] args) {
        Adapter client = new Adapter(); //Client
        client.makePayment(300);
    }
}
